File: strands-ts/src/storage/in-memory-storage.ts
import type { Storage } from './storage.js'

import { namespace, normalizeKey, normalizePrefix } from './storage.js'

/**
 * In-memory {@link Storage} backend backed by a `Map`.
 *
 * Useful for testing and for serverless environments where disk access is unavailable.
 * Content does not survive process restarts — for persistence use {@link LocalFileStorage}
 * or {@link S3Storage}.
 *
 * This is a plain unbounded store with no eviction. Consumers that need eviction
 * (e.g. the ContextOffloader plugin) manage it themselves.
 *
 * Like the other shipped backends, keys are normalized via {@link normalizeKey}:
 * slash runs are collapsed, leading/trailing slashes are stripped, and `..`
 * segments are rejected.
 *
 * @example
 * ```typescript
 * const storage = new InMemoryStorage()
 * await storage.write('memory/notes.json', new TextEncoder().encode('[]'))
 * const bytes = await storage.read('memory/notes.json')
 * ```
 */
export class InMemoryStorage implements Storage {
  private readonly _store = new Map<string, Uint8Array>()

  /**
   * Stores `data` under `key`, overwriting any existing value.
   * Bytes are copied on write to prevent aliasing with the caller's buffer.
   *
   * @param key - Opaque string key identifying the value
   * @param data - Raw bytes to persist
   * @throws {@link StorageError} if the key is empty or contains `..` segments
   */
  async write(key: string, data: Uint8Array): Promise<void> {
    this._store.set(normalizeKey(key), data.slice())
  }

  /**
   * Retrieves the bytes previously stored under `key`.
   * Returns a copy to prevent aliasing with the internal buffer.
   *
   * @param key - The key to read
   * @returns The stored bytes, or `null` if no value exists for `key`
   * @throws {@link StorageError} if the key is empty or contains `..` segments
   */
  async read(key: string): Promise<Uint8Array | null> {
    const value = this._store.get(normalizeKey(key))
    if (value === undefined) return null
    return value.slice()
  }

  /**
   * Deletes the value stored under `key`. A no-op if the key does not exist.
   *
   * @param key - The key to delete
   * @throws {@link StorageError} if the key is empty or contains `..` segments
   */
  async delete(key: string): Promise<void> {
    this._store.delete(normalizeKey(key))
  }

  /**
   * Lists the keys whose names begin with `prefix`, sorted lexicographically.
   *
   * @param prefix - Key prefix to match. An empty string matches all keys.
   * @returns The matching keys, sorted ascending
   * @throws {@link StorageError} if the prefix contains `..` segments
   */
  async list(prefix: string): Promise<string[]> {
    const normalized = normalizePrefix(prefix)
    const keys: string[] = []
    for (const key of this._store.keys()) {
      if (key.startsWith(normalized)) keys.push(key)
    }
    return keys.sort()
  }

  /** Returns a prefixed view of this storage without mutating the original. */
  namespace(prefix: string): Storage {
    return namespace(this, prefix)
  }

  /**
   * Removes all stored entries. Useful for resetting state between tests.
   */
  clear(): void {
    this._store.clear()
  }
}

File: strands-py/src/strands/storage/in_memory_storage.py
"""In-memory storage implementation."""

from __future__ import annotations

import builtins
import threading

from .storage import _NamespacedStorage, _normalize_key, _normalize_prefix


class InMemoryStorage:
    """Map-backed storage for testing and short-lived processes.

    Content does not survive process restarts. The store is unbounded — consumers
    manage eviction themselves.

    Example:
        ```python
        from strands.storage import InMemoryStorage

        storage = InMemoryStorage()
        await storage.write("sessions/abc/state.json", b'{"messages": []}')
        data = await storage.read("sessions/abc/state.json")
        ```
    """

    def __init__(self) -> None:
        """Initialize an empty in-memory store."""
        self._store: dict[str, bytes] = {}
        self._lock = threading.Lock()

    async def write(self, key: str, data: bytes) -> None:
        """Store data under key, overwriting any existing value.

        Args:
            key: Opaque string key identifying the value.
            data: Raw bytes to persist.

        Raises:
            StorageError: If the key is invalid.
        """
        normalized = _normalize_key(key)
        with self._lock:
            self._store[normalized] = bytes(data)

    async def read(self, key: str) -> bytes | None:
        """Retrieve the bytes previously stored under key.

        Args:
            key: The key to read.

        Returns:
            The stored bytes, or None if no value exists for key.

        Raises:
            StorageError: If the key is invalid.
        """
        normalized = _normalize_key(key)
        with self._lock:
            value = self._store.get(normalized)
        return value

    async def delete(self, key: str) -> None:
        """Delete the value stored under key. A no-op if the key does not exist.

        Args:
            key: The key to delete.

        Raises:
            StorageError: If the key is invalid.
        """
        normalized = _normalize_key(key)
        with self._lock:
            self._store.pop(normalized, None)

    async def list(self, query: str = "") -> builtins.list[str]:
        """List keys matching the given prefix.

        Args:
            query: A prefix string to filter keys. Empty string matches all.

        Returns:
            Matching keys sorted ascending.

        Raises:
            StorageError: If the prefix is invalid.
        """
        prefix = _normalize_prefix(query)
        with self._lock:
            keys = sorted(k for k in self._store if k.startswith(prefix))
        return keys

    def namespace(self, prefix: str) -> _NamespacedStorage:
        """Return a view of this storage with all keys prefixed.

        Args:
            prefix: Prefix to prepend to all keys.

        Returns:
            A namespaced storage view.
        """
        return _NamespacedStorage(self, prefix)

    def clear(self) -> None:
        """Remove all stored entries."""
        with self._lock:
            self._store.clear()
